src/main/java/org/antlr/jetbrains/adapter/lexer/PsiElementTypeFactory.java

```diff
@@ -92,11 +100,16 @@ public TokenIElementType getEofElementType() {
     }
 
     @NotNull
-    private List<RuleIElementType> createRuleIElementTypes(Language language, String[] ruleNames) {
+    private List<RuleIElementType> createRuleIElementTypes(Language language, String[] ruleNames,
+            Collection<RuleIElementType> customRuleElementTypes) {
         List<RuleIElementType> result;
-        RuleIElementType[] elementTypes = new RuleIElementType[ruleNames.length];
+        RuleIElementType[] elementTypes = new RuleIElementTypeImpl[ruleNames.length];
         for (int i = 0; i < ruleNames.length; i++) {
-            elementTypes[i] = new RuleIElementType(i, ruleNames[i], language);
+            elementTypes[i] = new RuleIElementTypeImpl(i, ruleNames[i], language);
+        }
+        for (RuleIElementType customType : customRuleElementTypes) {
+            assert customType instanceof IElementType;
+            elementTypes[customType.getRuleIndex()] = customType;
```

Comment:
Where do we get rule indexes for custom elements?. Are they unique? If so, I think it makes sense to throw an exception if custom type overrides already registered element type.